mrmustard/lab_dev/bb.py

```diff
@@ -0,0 +1,422 @@
+# Copyright 2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Branch and bound algorithm for optimal contraction of a tensor network.
+"""
+
+from __future__ import annotations
+import random
+from queue import PriorityQueue
+from math import factorial
+import numpy as np
+from typing import Generator
+import networkx as nx
+from mrmustard.lab_dev.wires import Wires
+from mrmustard.lab_dev.circuit_components import CircuitComponent
+
+Edge = tuple[int, int]
+
+
+class GraphComponent:
+    r"""A lightweight "CircuitComponent" without the actual representation.
+    Basically a wrapper around Wires, so that it can emulate components in
+    a circuit. It exposes the repr, wires, shape, name and cost of obtaining
+    the component from previous contractions.
+    """
+
+    def __init__(self, repr: str, wires: Wires, shape: list[int], name: str = "", cost: int = 0):
+        if None in shape:
+            raise ValueError("Detected `None`s in shape. Please provide a full shape.")
+        self.repr = repr
+        self.wires = wires
+        self.shape = list(shape)
+        self.name = name
+        self.cost = cost
+
+    @classmethod
+    def from_circuitcomponent(cls, c: CircuitComponent):
+        r"""
+        Creates a GraphComponent from a CircuitComponent.
+        """
+        return GraphComponent(
+            repr=str(c.representation.__class__.__name__),
+            wires=Wires(*c.wires.args),
+            shape=c.auto_shape(),
+            name=c.__class__.__name__,
+        )
+
+    def copy(self) -> GraphComponent:
+        """Returns a copy of the GraphComponent."""
+        return GraphComponent(self.repr, self.wires, self.shape, self.name)
```

Comment:
Is `cost` missing here intentional? Also do we need this? Shouldn't the builtin `copy`/`deepcopy` also work?

Reply:
Not intentional. Yep, deepcopy did it.

Reply:
In that case should `copy` be removed and just use `deepcopy`?